### ISSUE TITLE
Planning for astroid `3.0`

### ISSUE THREAD

**Issue by cdce8p** [pylint-dev/astroid#1868]
With the breaking change for the new `Try` node we should probably start planning for version `3.0`. We might as well use the major version bump to bundle other breaking changes as well.

## How to update pylint
With the tight coupling between pylint and astroid, one of the main challenges will be how to keep the changes to pylint manageable. After each major change, we could
* Release a new alpha version of astroid
* Or pin a specific astroid commit in pylint directly.
```ini
astroid @ git+https://github.com/cdce8p/astroid.git@<commit>
```

Either way would help us keep the pylint changes reviewable.

Any bugfix release would only be merged in the pylint maintenance branch.

## ToDo's
- [x] New `Try` node
https://github.com/PyCQA/astroid/pull/1867
- [ ] Redo Import nodes to better align with current AST -> Add `Alias` node
https://github.com/PyCQA/astroid/pull/912
https://github.com/PyCQA/astroid/issues/1339
- [x] Fully remove deprecated nodes -> `Ellipsis`, `Index`, `ExtSlice`
#2152
- [x] Remove deprecated default argument values. Mostly `None` for arguments which should never be `None`. That will help simplify typing.
- [x] Remove deprecated `doc` attributes
#2154
- [ ] Documentation for all breaking changes -> upgrade guide

## Maybe?
- [x] Better typing for `Uninferable`. Just an idea, with a breaking change we could change the class to `UninferableType` and `Uninferable` being an instance of it.
https://github.com/PyCQA/astroid/issues/1680
- [ ] Make all arguments to `__init__` and `postinit` for nodes keyword only args.
- [ ] Remove explicit node exports from `astroid.__init__`.
- [ ] Remove `astroid.scoped_nodes` module (not the package)

--
Those are just some things I had in mind. Happy to add additional things to it.
/CC: @Pierre-Sassoulas, @DanielNoord, @jacobtylerwalls

## Open Questions
Until we're done with all changes, I don't think it make sense to release a new minor version. Should we release `2.13` before we start the process just to get the last features out there? Any bug fixes could be backported as usual. The breaking changes are mostly for downstream libraries, not astroid itself.

**Comment by Pierre-Sassoulas** [comment]
>  Should we release 2.13 before we start the process just to get the last features out there? Any bug fixes could be backported as usual. The breaking changes are mostly for downstream libraries, not astroid itself.

Agree with that, the 2.12 maintenance was very long already. I think we waited for #1189 but it can be merged when it's ready.

**Comment by DanielNoord** [comment]
TODO list seems good to me. I'd like to help work on stuff, but I think we should try and do some concentrated efforts as that avoid the risks of many PRs ending up in review limbo.

**Comment by cdce8p** [comment]
> > Should we release 2.13 before we start the process just to get the last features out there? Any bug fixes could be backported as usual. The breaking changes are mostly for downstream libraries, not astroid itself.
> 
> Agree with that, the 2.12 maintenance was very long already. I think we waited for #1189 but it can be merged when it's ready. (Edit: #1189 seems to be ready for review, and already 2 persons approved, we should probably add this one in the TODO list.)

I'll try to take a look at it soon ™️ If it's good, it could make sense to include it with `2.13` before we start with merging `3.0` PRs.

> TODO list seems good to me. I'd like to help work on stuff, but I think we should try and do some concentrated efforts as that avoid the risks of many PRs ending up in review limbo.

Agreed, I opened the issue since I wanted to start the discussion early (and had the `Try` change ready). Ideally, we don't wait too long for the `3.0` release after we start merging breaking changes.

Fortunately, most of the changes are strait forward and easy to implement. The most challenging one left is probably the `Import` node rewrite. My idea was to prepare a PR beforehand, similarly to the one I already opened for `Try` before moving forward.

Since the overlap between the changes is minimal, that can be done for some of the other ones too, although not strictly necessary.

> I'd like to help work on stuff

These ones only require some time
* Removal of `Ellipsis`, `Index`, `ExtSlice`
* Remove deprecated `doc` attributes
* Remove explicit node exports from `astroid.__init__` (if we want to do it)

There is also the separate docs page, i.e. some kind of upgrade guide. Laying the foundation for it would help (create a file, add link to it from the main page, etc.). After that, everyone can start contributing to it much more easily.

**Comment by DanielNoord** [comment]
> There is also the separate docs page, i.e. some kind of upgrade guide. Laying the foundation for it would help (create a file, add link to it from the main page, etc.). After that, everyone can start contributing to it much more easily.

https://github.com/PyCQA/astroid/pull/1873 :smile: 

Least I could do!

**Comment by Pierre-Sassoulas** [comment]
I just released 2.12.13, I think it should be the last patch version based on 2.12. I'm wondering if we should release 2.13.0 without #1189. Marc reviewed and there's a lot of comments. We can always release 2.14.0 fast if the situation evolve.

**Comment by jacobtylerwalls** [comment]
The downside is that it's blocking several pylint PRs, and the unresolved comments are almost entirely cosmetic/small.

**Comment by DanielNoord** [comment]
If @cdce8p is interested in this I would be happy to do some cooperation on that PR and implement his comments in a back-and-forth process to just merge this somewhere this week.

I have some free time in GMT+1 evening this week so if that would work for you I think we can help move that PR along.

**Comment by cdce8p** [comment]
> If @cdce8p is interested in this I would be happy to do some cooperation on that PR and implement his comments in a back-and-forth process to just merge this somewhere this week.
> 
> I have some free time in GMT+1 evening this week so if that would work for you I think we can help move that PR along.

I will try to make some time, although I can't promise it.

--
As for the general timeframe for `3.0`. There are still several big todos I would like to take a look at before moving forward. So I think probably at least a month, maybe around New Year.

**Comment by jacobtylerwalls** [comment]
Duplicate of #2148, the two issues already linked.
